analytic_engine/src/instance/serializer.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    sync::{Arc, Condvar, Mutex},
+    time::Instant,
+};
+
+use common_util::{runtime::Runtime, time::InstantExt};
+use futures::Future;
+use log::error;
+use table_engine::table::{Error as TableError, Result as TableResult, TableId};
+use tokio::sync::oneshot;
+
+use crate::{
+    instance::flush_compaction::{self, BackgroundFlushFailed},
+    table::metrics::Metrics,
+};
+
+#[derive(Default)]
+enum FlushState {
+    #[default]
+    Ok,
+    Flushing,
+    Failed {
+        err_msg: String,
+    },
+}
+
+type ScheduleLock = Arc<(Mutex<FlushState>, Condvar)>;
+
+#[derive(Default)]
+pub struct TableFlushScheduler {
+    schedule_lock: ScheduleLock,
+}
+
+/// To ensure the consistency of a table's data, these rules are required:
+/// - The write procedure (write wal + write memtable) should be serialized as a
+///   whole, that is to say, it is not allowed to write wal and memtable
+///   concurrently or interleave the two sub-procedures;
+/// - Any operation that may change the data of a table should be serialized,
+///   including altering table schema, dropping table, etc;
+/// - The flush procedure of a table should be serialized;
+pub struct TableOpSerializer {
+    table_id: TableId,
+    flush_scheduler: TableFlushScheduler,
+}
+
+impl TableOpSerializer {
+    pub fn new(table_id: TableId) -> Self {
+        Self {
+            table_id,
+            flush_scheduler: TableFlushScheduler::default(),
+        }
+    }
+
+    #[inline]
+    pub fn table_id(&self) -> TableId {
+        self.table_id
+    }
+}
+
+impl TableOpSerializer {
+    pub fn flush_scheduler(&mut self) -> &mut TableFlushScheduler {
+        &mut self.flush_scheduler
+    }
+}
+
+impl TableFlushScheduler {
+    /// Control the flush procedure and ensure multiple flush procedures to be
+    /// sequential.
+    ///
+    /// REQUIRE: should only be called by the write thread.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn flush_sequentially<F, T>(
+        &mut self,
+        table: String,
+        metrics: &Metrics,
+        flush_job: F,
+        on_flush_success: T,
+        block_on_write_thread: bool,
+        runtime: &Runtime,
+        res_sender: Option<oneshot::Sender<TableResult<()>>>,
+    ) -> flush_compaction::Result<()>
+    where
+        F: Future<Output = flush_compaction::Result<()>> + Send + 'static,
+        T: Future<Output = ()> + Send + 'static,
+    {
+        // If flush operation is running, then we need to wait for it to complete first.
+        // Actually, the loop waiting ensures the multiple flush procedures to be
+        // sequential, that is to say, at most one flush is being executed at
+        // the same time.
+        {
+            let mut stall_begin: Option<Instant> = None;
+            let mut flush_state = self.schedule_lock.0.lock().unwrap();
```

Comment:
Maybe tokio::Mutex is more suitable here?
Because, we need to wait io of last flushing one?

Reply:
The lock won't come across the await point and is only be held for a while, so let's keep using the std::sync::Mutex. However, the CondVar may block the current thread which is unacceptable for tokio runtime.